### PR TITLE
Add writer routes

### PR DESCRIPTION
**Pull request by BrandonMurnane**

### CODE REVIEW

**Comment by kriscfoster** [comment]
@BrandonMurnane did you mean to check in changes to: package-lock.json, package.json?

**Comment by BrandonMurnane** [comment]
@iamigo for changes to package.json when someone installs could we change it to ~ rather than ^ to only pick up minor patches or say which version we want to stop this happening ?

File: api/v1/controllers/bots.js [hunk 1]
```diff
@@ -203,4 +208,72 @@ module.exports = {
...
+   *
+   * Retrieves all the writers associated with the bot
+   *
+   * @param {IncomingMessage} req - The request object
+   * @param {ServerResponse} res - The response object
+   * @param {Function} next - The next middleware function in the stack
+   */
+  getBotWriters(req, res, next) {
+    doGetWriters.getWriters(req, res, next, helper);
+  }, // getSubjectWriters
```
Comment (kriscfoster) [inline api/v1/controllers/bots.js]: getBotWriters

File: api/v1/controllers/bots.js [hunk 2]
```diff
@@ -203,4 +208,72 @@ module.exports = {
...
+   * Determine whether a user is an authorized writer for a bot and returns
+   * the user record if so.
+   *
+   * @param {IncomingMessage} req - The request object
+   * @param {ServerResponse} res - The response object
+   * @param {Function} next - The next middleware function in the stack
+   */
+  getBotWriter(req, res, next) {
+    doGetWriters.getWriter(req, res, next, helper);
+  }, // getSubjectWriter
```
Comment (kriscfoster) [inline api/v1/controllers/bots.js]: getBotWriter

File: api/v1/controllers/bots.js [hunk 3]
```diff
@@ -203,4 +208,72 @@ module.exports = {
...
+
+  /**
+   * POST /bots/{key}/writers
+   *
+   * Add one or more users to an bots list of authorized writers
```
Comment (kriscfoster) [inline api/v1/controllers/bots.js]: to a bot's list

File: api/v1/controllers/bots.js [hunk 4]
```diff
@@ -203,4 +208,72 @@ module.exports = {
...
+   *
+   * @param {IncomingMessage} req - The request object
+   * @param {ServerResponse} res - The response object
+   * @param {Function} next - The next middleware function in the stack
+   */
+  postBotWriters(req, res, next) {
+    doPostWriters(req, res, next, helper);
+  }, // postSubjectWriters
```
Comment (kriscfoster) [inline api/v1/controllers/bots.js]: postBotWriters

File: api/v1/controllers/roomTypes.js [hunk 5]
```diff
@@ -86,4 +90,72 @@ module.exports = {
     doPost(req, res, next, helper);
   },
 
+  /**
+   * GET /rooms/{key}/writers
+   *
+   * Retrieves all the writers associated with the room
```
Comment (kriscfoster) [inline api/v1/controllers/roomTypes.js]: with the roomType

File: api/v1/controllers/roomTypes.js [hunk 6]
```diff
@@ -86,4 +90,72 @@ module.exports = {
     doPost(req, res, next, helper);
   },
 
+  /**
+   * GET /rooms/{key}/writers
```
Comment (kriscfoster) [inline api/v1/controllers/roomTypes.js]: roomTypes/{key}/writers

File: api/v1/controllers/roomTypes.js [hunk 7]
```diff
@@ -86,4 +90,72 @@ module.exports = {
...
+   *
+   * Retrieves all the writers associated with the room
+   *
+   * @param {IncomingMessage} req - The request object
+   * @param {ServerResponse} res - The response object
+   * @param {Function} next - The next middleware function in the stack
+   */
+  getRoomTypeWriters(req, res, next) {
+    doGetWriters.getWriters(req, res, next, helper);
+  }, // getSubjectWriters
```
Comment (kriscfoster) [inline api/v1/controllers/roomTypes.js]: getRoomTypeWriters

File: api/v1/controllers/roomTypes.js [hunk 8]
```diff
@@ -86,4 +90,72 @@ module.exports = {
...
+
+  /**
+   * GET /rooms/{key}/writers/userNameOrId
```
Comment (kriscfoster) [inline api/v1/controllers/roomTypes.js]: roomTypes

File: api/v1/controllers/roomTypes.js [hunk 9]
```diff
@@ -86,4 +90,72 @@ module.exports = {
...
+   * Determine whether a user is an authorized writer for a room and returns
+   * the user record if so.
+   *
+   * @param {IncomingMessage} req - The request object
+   * @param {ServerResponse} res - The response object
+   * @param {Function} next - The next middleware function in the stack
+   */
+  getRoomTypeWriter(req, res, next) {
+    doGetWriters.getWriter(req, res, next, helper);
+  }, // getSubjectWriter
```
Comment (kriscfoster) [inline api/v1/controllers/roomTypes.js]: getRoomTypeWriter

File: api/v1/controllers/roomTypes.js [hunk 10]
```diff
@@ -86,4 +90,72 @@ module.exports = {
...
+
+  /**
+   * POST /rooms/{key}/writers
```
Comment (kriscfoster) [inline api/v1/controllers/roomTypes.js]: roomTypes

File: api/v1/controllers/roomTypes.js [hunk 11]
```diff
@@ -86,4 +90,72 @@ module.exports = {
...
+   *
+   * Add one or more users to an rooms list of authorized writers
+   *
+   * @param {IncomingMessage} req - The request object
+   * @param {ServerResponse} res - The response object
+   * @param {Function} next - The next middleware function in the stack
+   */
+  postRoomTypeWriters(req, res, next) {
+    doPostWriters(req, res, next, helper);
+  }, // postSubjectWriters
```
Comment (kriscfoster) [inline api/v1/controllers/roomTypes.js]: postRoomTypeWriters

File: api/v1/controllers/roomTypes.js [hunk 12]
```diff
@@ -86,4 +90,72 @@ module.exports = {
...
+   * @param {Function} next - The next middleware function in the stack
+   */
+  getRoomTypeWriter(req, res, next) {
+    doGetWriters.getWriter(req, res, next, helper);
+  }, // getSubjectWriter
+
+  /**
+   * POST /rooms/{key}/writers
+   *
+   * Add one or more users to an rooms list of authorized writers
```
Comment (kriscfoster) [inline api/v1/controllers/roomTypes.js]: roomTypes list of

File: api/v1/controllers/roomTypes.js [hunk 13]
```diff
@@ -86,4 +90,72 @@ module.exports = {
...
+   * @param {IncomingMessage} req - The request object
+   * @param {ServerResponse} res - The response object
+   * @param {Function} next - The next middleware function in the stack
+   */
+  postRoomTypeWriters(req, res, next) {
+    doPostWriters(req, res, next, helper);
+  }, // postSubjectWriters
+
+  /**
+   * DELETE /rooms/{keys}/writers
```
Comment (kriscfoster) [inline api/v1/controllers/roomTypes.js]: /roomTypes/

File: api/v1/controllers/rooms.js [hunk 14]
```diff
@@ -86,4 +90,72 @@ module.exports = {
...
+   *
+   * Retrieves all the writers associated with the room
+   *
+   * @param {IncomingMessage} req - The request object
+   * @param {ServerResponse} res - The response object
+   * @param {Function} next - The next middleware function in the stack
+   */
+  getRoomWriters(req, res, next) {
+    doGetWriters.getWriters(req, res, next, helper);
+  }, // getSubjectWriters
```
Comment (kriscfoster) [inline api/v1/controllers/rooms.js]: getRoomWriters

File: api/v1/controllers/rooms.js [hunk 15]
```diff
@@ -86,4 +90,72 @@ module.exports = {
...
+   * Determine whether a user is an authorized writer for a room and returns
+   * the user record if so.
+   *
+   * @param {IncomingMessage} req - The request object
+   * @param {ServerResponse} res - The response object
+   * @param {Function} next - The next middleware function in the stack
+   */
+  getRoomWriter(req, res, next) {
+    doGetWriters.getWriter(req, res, next, helper);
+  }, // getSubjectWriter
```
Comment (kriscfoster) [inline api/v1/controllers/rooms.js]: getRoomWriter

File: api/v1/controllers/rooms.js [hunk 16]
```diff
@@ -86,4 +90,72 @@ module.exports = {
...
+   *
+   * Add one or more users to an rooms list of authorized writers
+   *
+   * @param {IncomingMessage} req - The request object
+   * @param {ServerResponse} res - The response object
+   * @param {Function} next - The next middleware function in the stack
+   */
+  postRoomWriters(req, res, next) {
+    doPostWriters(req, res, next, helper);
+  }, // postSubjectWriters
```
Comment (kriscfoster) [inline api/v1/controllers/rooms.js]: postRoomWriters

File: db/model/bot.js [hunk 17]
```diff
@@ -168,5 +168,19 @@ module.exports = function bot(seq, dataTypes) {
     });
   };
 
+  Bot.prototype.isWritableBy = function (who) {
```
Comment (cassiodias) [inline db/model/bot.js]: Is there a better way to reuse this fn? its the same ofr botAction, botData, room and roomType

**Comment by iamigo** [comment]
@BrandonMurnane re: changes to package.json
any module in particular you're thinking about here? what changes are you trying to prevent?

File: tests/api/v1/botActions/deleteWithoutPerms.js [hunk 18]
```diff
@@ -0,0 +1,103 @@
...
+const pfx = '___';
+
+describe('tests/api/v1/botActions/deleteWithoutPerms.js >', () => {
+  let testBotAction;
+  let validToken;
+  let invalidToken;
+  let user;
+
+  before((done) => {
+    tu.createToken()
```
Comment (kriscfoster) [inline tests/api/v1/botActions/deleteWithoutPerms.js]: I don't think you need this as you are creating a token below for the user. This token is never used

File: tests/api/v1/botActions/deleteWithoutPerms.js [hunk 19]
```diff
@@ -0,0 +1,103 @@
...
+    .end(done);
+  });
+
+  it('deleting BotAction with permission should return 200', (done) => {
+    api.delete(`${path}/${testBotAction.id}`)
+    .set('Authorization', validToken)
+    .expect(constants.httpStatus.OK)
+    .end(done);
+  });
+
```
Comment (kriscfoster) [inline tests/api/v1/botActions/deleteWithoutPerms.js]: no newline needed

File: tests/api/v1/botActions/patchWithoutPerms.js [hunk 20]
```diff
@@ -0,0 +1,109 @@
...
+const pfx = '___';
+
+describe('tests/api/v1/botActions/patchWithoutPerms.js >', () => {
+  let testBotAction;
+  let validToken;
+  let invalidToken;
+  let user;
+
+  before((done) => {
+    tu.createToken()
```
Comment (kriscfoster) [inline tests/api/v1/botActions/patchWithoutPerms.js]: Same as above, dont think you need this line

File: tests/api/v1/botActions/patchWithoutPerms.js [hunk 21]
```diff
@@ -0,0 +1,109 @@
...
+  it('deleting BotAction with permission should return 200', (done) => {
+    api.patch(`${path}/${testBotAction.id}`)
+    .set('Authorization', validToken)
+    .send({
+      isPending: false,
+    })
+    .expect(constants.httpStatus.OK)
+    .end(done);
+  });
+
```
Comment (kriscfoster) [inline tests/api/v1/botActions/patchWithoutPerms.js]: no newline needed

File: tests/api/v1/botData/deleteWithoutPerms.js [hunk 22]
```diff
@@ -0,0 +1,103 @@
...
+const pfx = '___';
+
+describe('tests/api/v1/botData/deleteWithoutPerms.js >', () => {
+  let testBotData;
+  let validToken;
+  let invalidToken;
+  let user;
+
+  before((done) => {
+    tu.createToken()
```
Comment (kriscfoster) [inline tests/api/v1/botData/deleteWithoutPerms.js]: Same as above, don't think you need this

File: tests/api/v1/botData/deleteWithoutPerms.js [hunk 23]
```diff
@@ -0,0 +1,103 @@
...
+    .end(done);
+  });
+
+  it('deleting BotData with permission should return 200', (done) => {
+    api.delete(`${path}/${testBotData.id}`)
+    .set('Authorization', validToken)
+    .expect(constants.httpStatus.OK)
+    .end(done);
+  });
+
```
Comment (kriscfoster) [inline tests/api/v1/botData/deleteWithoutPerms.js]: no newline needed

File: tests/api/v1/botData/patchWithoutPerms.js [hunk 24]
```diff
@@ -0,0 +1,109 @@
...
+  it('deleting BotData with permission should return 200', (done) => {
+    api.patch(`${path}/${testBotData.id}`)
+    .set('Authorization', validToken)
+    .send({
+      value: 'apples',
+    })
+    .expect(constants.httpStatus.OK)
+    .end(done);
+  });
+
```
Comment (kriscfoster) [inline tests/api/v1/botData/patchWithoutPerms.js]: no newline needed

File: tests/api/v1/bots/deleteWithoutPerms.js [hunk 25]
```diff
@@ -0,0 +1,77 @@
...
+const pfx = '___';
+
+describe('tests/api/v1/bots/patchWithoutPerms.js >', () => {
+  let testBot;
+  let validToken;
+  let invalidToken;
+  let user;
+
+  before((done) => {
+    tu.createToken()
```
Comment (kriscfoster) [inline tests/api/v1/bots/deleteWithoutPerms.js]: don't think you need this

File: tests/api/v1/bots/deleteWithoutPerms.js [hunk 26]
```diff
@@ -0,0 +1,77 @@
...
+    .end(done);
+  });
+
+  it('deleting roomType with permission should return 200', (done) => {
+    api.delete(`${path}/${testBot.id}`)
+    .set('Authorization', validToken)
+    .expect(constants.httpStatus.OK)
+    .end(done);
+  });
+
```
Comment (kriscfoster) [inline tests/api/v1/bots/deleteWithoutPerms.js]: no newline needed

File: tests/api/v1/bots/patchWithoutPerms.js [hunk 27]
```diff
@@ -0,0 +1,83 @@
...
+  it('deleting roomType with permission should return 200', (done) => {
+    api.patch(`${path}/${testBot.id}`)
+    .set('Authorization', validToken)
+    .send({
+      version: '2.0',
+    })
+    .expect(constants.httpStatus.OK)
+    .end(done);
+  });
+
```
Comment (kriscfoster) [inline tests/api/v1/bots/patchWithoutPerms.js]: no newline needed

File: tests/api/v1/roomTypes/deleteWithoutPerms.js [hunk 28]
```diff
@@ -0,0 +1,77 @@
...
+const pfx = '___';
+
+describe('tests/api/v1/roomTypes/deleteWithoutPerms.js >', () => {
+  let testRoomType;
+  let validToken;
+  let invalidToken;
+  let user;
+
+  before((done) => {
+    tu.createToken()
```
Comment (kriscfoster) [inline tests/api/v1/roomTypes/deleteWithoutPerms.js]: same as above

File: tests/api/v1/roomTypes/deleteWithoutPerms.js [hunk 29]
```diff
@@ -0,0 +1,77 @@
...
+    .end(done);
+  });
+
+  it('deleting roomType with permission should return 200', (done) => {
+    api.delete(`${path}/${testRoomType.id}`)
+    .set('Authorization', validToken)
+    .expect(constants.httpStatus.OK)
+    .end(done);
+  });
+
```
Comment (kriscfoster) [inline tests/api/v1/roomTypes/deleteWithoutPerms.js]: no newline needed

File: tests/api/v1/roomTypes/patchWithoutPerms.js [hunk 30]
```diff
@@ -0,0 +1,83 @@
...
+  it('deleting roomType with permission should return 200', (done) => {
+    api.patch(`${path}/${testRoomType.id}`)
+    .set('Authorization', validToken)
+    .send({
+      version: '2.0',
+    })
+    .expect(constants.httpStatus.OK)
+    .end(done);
+  });
+
```
Comment (kriscfoster) [inline tests/api/v1/roomTypes/patchWithoutPerms.js]: same as above

File: tests/api/v1/roomTypes/patchWithoutPerms.js [hunk 31]
```diff
@@ -0,0 +1,83 @@
...
+const pfx = '___';
+
+describe('tests/api/v1/roomTypes/patchWithoutPerms.js >', () => {
+  let testRoomType;
+  let validToken;
+  let invalidToken;
+  let user;
+
+  before((done) => {
+    tu.createToken()
```
Comment (kriscfoster) [inline tests/api/v1/roomTypes/patchWithoutPerms.js]: same as above

File: tests/api/v1/rooms/deleteWithoutPerms.js [hunk 32]
```diff
@@ -0,0 +1,90 @@
...
+const pfx = '___';
+
+describe('tests/api/v1/rooms/deleteWithoutPerms.js >', () => {
+  let testRoom;
+  let validToken;
+  let invalidToken;
+  let user;
+
+  before((done) => {
+    tu.createToken()
```
Comment (kriscfoster) [inline tests/api/v1/rooms/deleteWithoutPerms.js]: same as above

File: tests/api/v1/rooms/deleteWithoutPerms.js [hunk 33]
```diff
@@ -0,0 +1,90 @@
...
+    .end(done);
+  });
+
+  it('deleting room with permission should return 200', (done) => {
+    api.delete(`${path}/${testRoom.id}`)
+    .set('Authorization', validToken)
+    .expect(constants.httpStatus.OK)
+    .end(done);
+  });
+
```
Comment (kriscfoster) [inline tests/api/v1/rooms/deleteWithoutPerms.js]: same as above

File: tests/api/v1/rooms/patchWithoutPerms.js [hunk 34]
```diff
@@ -0,0 +1,96 @@
...
+const pfx = '___';
+
+describe('tests/api/v1/rooms/deleteWithoutPerms.js >', () => {
+  let testRoom;
+  let validToken;
+  let invalidToken;
+  let user;
+
+  before((done) => {
+    tu.createToken()
```
Comment (kriscfoster) [inline tests/api/v1/rooms/patchWithoutPerms.js]: same as above

File: tests/api/v1/rooms/patchWithoutPerms.js [hunk 35]
```diff
@@ -0,0 +1,96 @@
...
+  it('deleting room with permission should return 200', (done) => {
+    api.patch(`${path}/${testRoom.id}`)
+    .set('Authorization', validToken)
+    .send({
+      active: false,
+    })
+    .expect(constants.httpStatus.OK)
+    .end(done);
+  });
+
```
Comment (kriscfoster) [inline tests/api/v1/rooms/patchWithoutPerms.js]: same as above

File: db/model/bot.js (hunk 17 above)
Comment (BrandonMurnane) [reply db/model/bot.js]: i had tried to do this but kept running into problems with it missing context

**Comment by tausifmuzaffar** [comment]
@cassiodias @iamigo are there still concerns. Since @kriscfoster is out should we remove him from the review @BrandonMurnane ?

File: db/model/bot.js (hunk 17 above)
Comment (cassiodias) [reply db/model/bot.js]: Ok! Maybe we could create a spike to look into that?

@tausifmuzaffar

**Pull request merged**